src/NHibernate/Util/LinkHashMap.cs

```diff
@@ -0,0 +1,641 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.Serialization;
+using System.Text;
+using NHibernate.DebugHelpers;
+
+namespace NHibernate.Util
+{
+	/// <summary>
+	/// A map of objects whose mapping entries are sequenced based on the order in which they were
+	/// added. This data structure has fast <c>O(1)</c> search time, deletion time, and insertion time
+	/// </summary>
+	/// <remarks>
+	/// This class is not thread safe.
+	/// This class is not really a replication of JDK LinkedHashMap{K, V}, 
+	/// this class is an adaptation of SequencedHashMap with generics.
+	/// </remarks>
+	[DebuggerTypeProxy(typeof(CollectionProxy<>))]
+	[Serializable]
+	internal class LinkHashMap<TKey, TValue> : IDictionary<TKey, TValue>, IDeserializationCallback
+	{
+		[Serializable]
+		protected class Entry
+		{
+			private readonly TKey _key;
+			private TValue _evalue;
```

Comment:
The `e` was very likely a "trick" to dodge the fact that `value` is a reserved word. (Instead of using `@` which is meant for this purpose.) There are no reason to keep it.

```suggestion
			private TValue _value;
```

Reply:
Maybe, I wasn't sure because of the serialisation...
Also one question bother me: may I
- remove fields by using auto properties
- or remove properties and make fields "public"?
what would be a safe move to clean `Entry` class?

Reply:
You can remove fields by using auto properties.

Reply:
OK, I've renamed `_evalue` to `_value`, but I did not use auto properties because generated backend fields have longer names and I realised that size of binary serialized object (stream/array) depends on:
- length of field names, 1 byte per character
- length of parent class name for nested classes, 3 bytes per character
- length of class name, 3 bytes per character
- length of namespace name, 3 bytes per character

Current size of 1 serialized object like `LinkHashMap<int,int>.Entry` is 826 bytes long.
- Using of auto properties makes it 890 bytes long (+7.75%)
- Extracting `Entry` class directly to NHibernate namespace and using field names like:
```c#
private readonly TKey k;
private TValue v;
private Entry<TKey,TValue> n;
private Entry<TKey,TValue> p;
```
makes it 759 bytes long (-8.11%)

So, I would not use auto properties and also consider to rename fields to 1 character long names at least. This will make it 810 bytes long (-1.94%).

Any thought? Because I don't really know how important it is (I don't use serialisation in NHibernate and not familiar with this feature, but it seems important to mention).

Reply:
We do not really care about the resulting size of the serialized binary. I have no recollection of this subject having ever been raised. Binary serialization is for us a legacy that we do not want to break in a minor: as long as it keeps working, that is good enough for us.

So, no need to bother about the resulting size. Code cleanliness is more important for us than the impacts it may have on the size of state binary serialized. (No one letter backing fields please!)